build/result.go

```diff
@@ -388,7 +388,7 @@ func populateProcessConfigFromResult(req *gateway.StartRequest, res *gateway.Res
 	} else if img != nil {
 		args = append(args, img.Config.Entrypoint...)
 	}
-	if cfg.Cmd != nil {
+	if !cfg.NoCmd {
```

Comment:
Just wondering, would it be possible to just change this to `len(cfg.Cmd) != 0` for a simpler fix?

Can a user actually specify an empty cmd? I'm not sure if this has a parallel with `docker run` (cc @tianon)

Reply:
This logic was hard for me to reason about, so I'm going to give some examples of `docker run` behavior in the hopes that it helps make things more clear. :sweat_smile:

Given an image with `ENTRYPOINT` set to `a` and `CMD` set to `b` (keeping these intentionally simple and including the obvious ones too -- let's not get into shell form vs JSON form because that gets nasty and doesn't really change this basic operation very much :joy:):

- `docker run img` -> runs `a b`
- `docker run img c` -> runs `a c`
- `docker run --entrypoint c img` -> runs `c` (no `a`, but more importantly no `b` also!)
- `docker run --entrypoint c img d` -> `c d`

In other words, the behavior is similar to that of `ENTRYPOINT` within a `Dockerfile` - if it's specified, it resets `CMD` (so we have to detect command from the CLI separate from command from the image).

Reply:
Thanks for the explanation! Based on the examples, `len(cfg.Cmd) != 0` doesn't look like a solution here because it doesn't implement the behaviour of `docker run --entrypoint c img` (`--invoke entrypoint=c` on buildx) that specifies an empty cmd and needs to ignore the image's default.